core/src/main/scala/org/apache/spark/executor/ProcfsMetricsGetter.scala

```diff
@@ -0,0 +1,231 @@
+  var isAvailable: Boolean = isProcfsAvailable
+  private val pid = computePid()
+
+  private lazy val isProcfsAvailable: Boolean = {
+    if (testing) {
+       true
+    }
+    else {
+      val procDirExists = Try(Files.exists(Paths.get(procfsDir))).recover {
+        case ioe: IOException =>
+          logWarning("Exception checking for procfs dir", ioe)
+          false
+      }
+      val shouldLogStageExecutorMetrics =
+        SparkEnv.get.conf.get(config.EVENT_LOG_STAGE_EXECUTOR_METRICS)
+      val shouldLogStageExecutorProcessTreeMetrics =
+        SparkEnv.get.conf.get(config.EVENT_LOG_PROCESS_TREE_METRICS)
+      procDirExists.get && shouldLogStageExecutorProcessTreeMetrics && shouldLogStageExecutorMetrics
+    }
+  }
+
+  private def computePid(): Int = {
+    if (!isAvailable || testing) {
+      return -1;
+    }
+    try {
+      // This can be simplified in java9:
+      // https://docs.oracle.com/javase/9/docs/api/java/lang/ProcessHandle.html
+      val cmd = Array("bash", "-c", "echo $PPID")
+      val out = Utils.executeAndGetOutput(cmd)
+      val pid = Integer.parseInt(out.split("\n")(0))
+      return pid;
+    }
+    catch {
+      case e: SparkException =>
+        logWarning("Exception when trying to compute process tree." +
+          " As a result reporting of ProcessTree metrics is stopped", e)
+        isAvailable = false
+        -1
+    }
+  }
+
+  private def computePageSize(): Long = {
+    if (testing) {
+      return pSizeForTest;
+    }
+    try {
+      val cmd = Array("getconf", "PAGESIZE")
+      val out = Utils.executeAndGetOutput(cmd)
+      Integer.parseInt(out.split("\n")(0))
+    } catch {
+      case e: Exception =>
+        logWarning("Exception when trying to compute pagesize, as a" +
+          " result reporting of ProcessTree metrics is stopped")
+        isAvailable = false
+        0
+    }
+  }
+
+  private def computeProcessTree(): Set[Int] = {
+    if (!isAvailable || testing) {
+      return Set()
+    }
+    var ptree: Set[Int] = Set()
+    ptree += pid
+    val queue = mutable.Queue.empty[Int]
+    queue += pid
+    while( !queue.isEmpty ) {
+      val p = queue.dequeue()
+      val c = getChildPids(p)
+      if(!c.isEmpty) {
+        queue ++= c
+        ptree ++= c.toSet
+      }
+    }
+    ptree
+  }
+
+  private def getChildPids(pid: Int): ArrayBuffer[Int] = {
+    try {
+      val builder = new ProcessBuilder("pgrep", "-P", pid.toString)
+      val process = builder.start()
+      val childPidsInInt = mutable.ArrayBuffer.empty[Int]
+      def appendChildPid(s: String): Unit = {
+        if (s != "") {
+          logTrace("Found a child pid:" + s)
+          childPidsInInt += Integer.parseInt(s)
+        }
+      }
+      val stdoutThread = Utils.processStreamByLine("read stdout for pgrep",
+        process.getInputStream, appendChildPid)
+      val error = process.getErrorStream
+      var errorString = ""
+      (0 until error.available()).foreach { i =>
```

Comment:
unfortunately this isn't going to work in general.  `available()` is just going to tell you an estimate of how much input is available *right now*, not how much data will ever be available on that stream.  You really just have to read from that stream in another thread, and wait for the process to finish.  I think you can simplify this with `Utils.processStreamByLine`:

```scala
val errorStringBuilder = new StringBuilder()
val stdErrThread = Utils.processStreamLineByLine("stderr for pgrep", process.getErroStream, { line =>
  errorStringBuilder.append(line)
})
val exitCode = process.waitFor()
stdoutThread.join()
stderrThread.join()
val errorString = errorStringBuilder.toString()
```

Reply:
btw -- I realize I'm being somewhat paranoid here.  `Utils.executeAndGetOutput()` has this issue if `redirectStderr = false`.

try this with this script:

```sh
more lotsOfStderr.sh 
n=$1
for (( c=1; c<=n; c++ ))
do
  echo "stderr $c" 1>&2
  if [ $(( $c % 100 )) -eq 0 ] ; then
    echo "stdout $c"
  fi
done
```

and then in a spark shell

```
scala> val clz = Class.forName("org.apache.spark.util.Utils")
clz: Class[_] = class org.apache.spark.util.Utils

scala> val m = clz.getMethods().filter{_.getName() == "executeAndGetOutput"}.head
m: java.lang.reflect.Method = public static java.lang.String org.apache.spark.util.Utils.executeAndGetOutput(scala.collection.Seq,java.io.File,scala.collection.Map,boolean)

scala>  m.invoke(null, Seq("/Users/irashid/lotsOfStderr.sh"), new java.io.File("."), Map(), false.asInstanceOf[java.lang.Boolean])
 
[... lots of output]

scala>  m.invoke(null, Seq("/Users/irashid/lotsOfStderr.sh", "6000"), new java.io.File("."), Map(), false.asInstanceOf[java.lang.Boolean])

[blocks forever]
```

so ~5k lines of stderr seems fine, it gets buffered, but at 6K its stuck waiting for something to read from stderr.  That limit will probably vary by system.

you could just ignore stderr completely, and hope its under whatever the limit is ...

Reply:
actually, nobody is calling `executeAndGetOutput` with `redirectStderr = false`.  we should remove that option completely from that method since its not sound.

Reply:
Just for my own understanding, Utils.processStreamByLine doesn't redirect the stderror as far as I can see in the code. So should I do that and then ignore it?